solidity/test/Mailbox.t.sol

```diff
@@ -0,0 +1,413 @@
+            metadata
+        );
+        assertEq(quote, defaultFee + requiredFee);
+
+        expectHookQuote(requiredHook, metadata, message);
+        expectHookQuote(overrideHook, metadata, message);
+        quote = mailbox.quoteDispatch(
+            remoteDomain,
+            address(recipient).addressToBytes32(),
+            body,
+            metadata,
+            overrideHook
+        );
+        assertEq(quote, overrideFee + requiredFee);
+    }
+
+    event Dispatch(
+        address indexed sender,
+        uint32 indexed destination,
+        bytes32 indexed recipient,
+        bytes message
+    );
+
+    event DispatchId(bytes32 indexed messageId);
+
+    function expectDispatch(
+        TestHook firstHook,
+        TestHook hook,
+        bytes calldata metadata,
+        bytes calldata body
+    ) internal {
+        bytes memory message = mailbox.buildOutboundMessage(
+            remoteDomain,
+            recipientb32,
+            body
+        );
+        expectHookQuote(firstHook, metadata, message);
+        expectHookPost(firstHook, metadata, message, firstHook.fee());
+        expectHookPost(hook, metadata, message, hook.fee());
+        vm.expectEmit(true, true, true, true, address(mailbox));
+        emit Dispatch(address(this), remoteDomain, recipientb32, message);
+        vm.expectEmit(true, false, false, false, address(mailbox));
+        emit DispatchId(message.id());
+    }
+
+    function test_dispatch(
+        uint8 n,
+        bytes calldata body,
+        bytes calldata metadata
+    ) public {
+        bytes calldata defaultMetadata = metadata[0:0];
+        uint32 nonce;
+        bytes32 id;
+
+        for (uint256 i = 0; i < n; i += 3) {
+            nonce = mailbox.nonce();
+            assertEq(nonce, i);
+
+            // default hook and no metadata
+            expectDispatch(requiredHook, defaultHook, defaultMetadata, body);
+            id = mailbox.dispatch(remoteDomain, recipientb32, body);
+            assertEq(mailbox.latestDispatchedId(), id);
+            nonce = mailbox.nonce();
+            assertEq(nonce, i + 1);
+
+            // default hook with metadata
+            expectDispatch(requiredHook, defaultHook, metadata, body);
+            id = mailbox.dispatch(remoteDomain, recipientb32, body, metadata);
+            assertEq(mailbox.latestDispatchedId(), id);
+            nonce = mailbox.nonce();
+            assertEq(nonce, i + 2);
+
+            // override default hook with metadata
+            expectDispatch(requiredHook, overrideHook, metadata, body);
+            id = mailbox.dispatch(
+                remoteDomain,
+                recipientb32,
+                body,
+                metadata,
+                overrideHook
+            );
+            assertEq(mailbox.latestDispatchedId(), id);
+            nonce = mailbox.nonce();
+            assertEq(nonce, i + 3);
+        }
+    }
+
+    function test_100dispatch_withMerkleTreeHook(bytes calldata body) public {
```

Comment:
can fuzz 100 too?

Reply:
for sure
this was mostly for instrumenting gas

Reply:
added a comment